File: c/lib/std-dev.c
#include <strings.h>
#include <math.h>

void std_dev(
          double * restrict devs,
    const double * const    vals,
    const double * const    mas,
    const int               bars,
    const int               period
) {
    for (int bar = 0; bar < period - 1; ++bar)
        devs[bar] = 0.0;

    const double p_inv = 1.0 / period;
    double sum_sq = 0.0;
    for (int i = 0; i < period; ++i) {
        double valsI = vals[i];
        sum_sq += valsI * valsI;
    }
    double mean     = mas[period - 1];
    double mean_sq  = mean * mean;
    double variance = sum_sq * p_inv - mean_sq;
    devs[period - 1] = variance > 0.0 ? sqrt(variance) : 0.0;
    for (int i = period; i < bars; ++i) {
        double vals_i  = vals[i];
        double vals_ip = vals[i - period];
        sum_sq  += vals_i * vals_i - vals_ip * vals_ip;
        mean     = mas[i];
        mean_sq  = mean * mean;
        variance = sum_sq * p_inv - mean * mean;
        devs[i]  = variance > 0.0 ? sqrt(variance) : 0.0;
    }
}

File: fortran/lib/std-dev.f90
module std_dev_module
    use iso_fortran_env, only : real64
    implicit none
    private

    public std_dev

    contains

subroutine std_dev(devs, vals, mas, period)
    real(real64), intent(out) :: devs(:)
    real(real64), intent(in)  :: vals(:)
    real(real64), intent(in)  :: mas(:)
    integer,      intent(in)  :: period

    real(real64) :: sum, ma, d, p_inv
    integer      :: bar, n

    ! Input validation
    n = size(devs)
    if (period < 1 .or. period > size(vals) .or. n < size(mas) .or. size(vals) < size(mas)) then
        devs = 0.0_real64
        return
    end if

    ! Initialize output
    devs(1:min(period, n)) = 0.0_real64

    p_inv = 1.0_real64 / real(period, real64)

    ! Compute initial sum for the first valid window
    sum = 0.0_real64
    if (n > period) then
        ma = mas(period + 1)
        do bar = 1, period
            d = vals(period + 1 - bar) - ma
            sum = sum + d * d
        end do
        devs(period + 1) = sqrt(sum * p_inv)
    end if

    ! Slide the window for remaining bars
    do bar = period + 2, n
        ma = mas(bar)
        ! Remove oldest element
        d = vals(bar - period) - mas(bar - 1)
        sum = sum - d * d
        ! Add newest element
        d = vals(bar) - ma
        sum = sum + d * d
        ! Ensure sum doesn't go negative due to rounding
        if (sum < 0.0_real64) sum = 0.0_real64
        devs(bar) = sqrt(sum * p_inv)
    end do
end subroutine std_dev

end module std_dev_module
